ZipException: invalid entry compressed size error with Java < 16

The plugin seems to be reusing `JarEntry` entries when adding the automatic module name in `ExtraModuleInfoTransform`, and with Java versions prior to 16 this may lead to errors like the following:

```javastacktrace
Caused by: java.lang.RuntimeException: java.util.zip.ZipException: invalid entry compressed size (expected 1665 but got 1680 bytes)
        at de.jjohannes.gradle.javamodules.ExtraModuleInfoTransform.addAutomaticModuleName(ExtraModuleInfoTransform.java:137)
        at de.jjohannes.gradle.javamodules.ExtraModuleInfoTransform.transform(ExtraModuleInfoTransform.java:64)
        at org.gradle.api.internal.artifacts.transform.DefaultTransformer.transform(DefaultTransformer.java:264)
        at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvocationFactory$AbstractTransformerExecution$1.call(DefaultTransformerInvocationFactory.java:296)
        at org.gradle.api.internal.artifacts.transform.DefaultTransformerInvocationFactory$AbstractTransformerExecution$1.call(DefaultTransformerInvocationFactory.java:291)
```
The above exception trace was triggered by the `org.w3c.css:sac:1.3` artifact with `extra-java-module-info` version 0.10:
```groovy
extraJavaModuleInfo {
    automaticModule('sac-1.3.jar', 'sac')
}
```

## Issue context
From the [JDK bug 8253952](https://bugs.openjdk.java.net/browse/JDK-8253952):
> The correct way of copying all entries from one zip file into another requires the creation of a new `ZipEntry` or at least resetting of the compressed size field. E.g.:
```java
 while((entry = zis.getNextEntry()) != null) {
     ZipEntry newEntry = new ZipEntry(entry.getName());
     zos.putNextEntry(newEntry);
     zis.transferTo(zos);
 }
```
> or:
```java
 while((entry = zis.getNextEntry()) != null) {
     entry.setCompressedSize(-1);
     zos.putNextEntry(entry);
     zis.transferTo(zos);
 }
```
That JDK issue is very detailed and almost says it all, but there is also a blog post that covers the topic: [How to programmatically copy jar files](http://javahowto.blogspot.com/2011/07/how-to-programmatically-copy-jar-files.html).

## How to reproduce
To reproduce the error, clone the CSS4J repository and build the `1-stable` branch with Java 11:
```console
git clone https://github.com/css4j/css4j.git
cd css4j
git checkout 1-stable
export JAVA_HOME="C:\\Program Files\\Eclipse Adoptium\\jdk-11.0.14.9-hotspot\\" # or your path to Java 11
./gradlew build
```

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -1,10 +1,18 @@
 name: Build Plugin
-on: push
+on: [ push, pull_request ]
 jobs:
   build:
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        jdk: [ 11 ]
     steps:
       - uses: actions/checkout@v2
+      - uses: actions/setup-java@v2
+        with:
+          java-version: ${{ matrix.jdk }}
+          distribution: temurin
       - uses: eskatos/gradle-command-action@v1
         with:
           arguments: :build

--- a/src/functionalTest/groovy/de/jjohannes/gradle/javamodules/test/ExtraJavaModuleInfoTest.groovy
+++ b/src/functionalTest/groovy/de/jjohannes/gradle/javamodules/test/ExtraJavaModuleInfoTest.groovy
@@ -675,6 +675,58 @@ class ExtraJavaModuleInfoTest extends Specification {
         build().task(':compileJava').outcome == TaskOutcome.SUCCESS
     }
 
+    // See: https://github.com/jjohannes/extra-java-module-info/issues/27
+    def "correctly handles copying JAR contents on JDKs < 16"() {
+        given:
+        new File(testFolder.root, "src/main/java/org/gradle/sample/app/data").mkdirs()
+
+        testFolder.newFile("src/main/java/org/gradle/sample/app/Main.java") << """
+            package org.gradle.sample.app;
+            
+            public class Main {
+                public static void main(String[] args) {     
+                    System.out.println(org.w3c.css.sac.Parser.class);           
+                }
+            }
+        """
+
+        testFolder.newFile("src/main/java/module-info.java") << """
+            module org.gradle.sample.app {
+                requires sac;
+            }
+        """
+        testFolder.newFile("build.gradle.kts") << """
+            plugins {
+                application
+                id("de.jjohannes.extra-java-module-info")
+            }
+            
+            repositories {
+                mavenCentral()
+            }
+            
+            java {
+                modularity.inferModulePath.set(true)
+            }
+            
+            dependencies {
+                implementation("org.w3c.css:sac:1.3")    
+            }
+            
+            extraJavaModuleInfo {
+                automaticModule("sac-1.3.jar", "sac")
+            }
+            
+            application {
+                mainModule.set("org.gradle.sample.app")
+                mainClass.set("org.gradle.sample.app.Main")
+            }
+        """
+
+        expect:
+        build().task(':compileJava').outcome == TaskOutcome.SUCCESS
+    }
+
     BuildResult build() {
         gradleRunnerFor(['build']).build()
     }

--- a/src/main/java/de/jjohannes/gradle/javamodules/ExtraModuleInfoTransform.java
+++ b/src/main/java/de/jjohannes/gradle/javamodules/ExtraModuleInfoTransform.java
@@ -165,6 +165,7 @@ abstract public class ExtraModuleInfoTransform implements TransformAction<ExtraM
                 providers.put(entryName.substring(SERVICES_PREFIX.length()), extractImplementations(content));
             }
             if (!JAR_SIGNATURE_PATH.matcher(jarEntry.getName()).matches() && !"META-INF/MANIFEST.MF".equals(jarEntry.getName())) {
+                jarEntry.setCompressedSize(-1);
                 outputStream.putNextEntry(jarEntry);
                 outputStream.write(content);
                 outputStream.closeEntry();

